core/src/server/rpc_module.rs

```diff
@@ -754,6 +754,183 @@ impl<Context: Send + Sync + 'static> RpcModule<Context> {
 		Ok(callback)
 	}
 
+	/// Similar to [`RpcModule::register_async_subscription`] but a little lower-level API
+	/// where handling the subscription is managed the user i.e, polling the subscription
+	/// such as spawning a separate task to do so.
+	///
+	/// This is more efficient as this doesn't require cloning the `params` in the subscription
+	///
+	/// # Examples
+	///
+	/// ```no_run
+	///
+	/// use jsonrpsee_core::server::{RpcModule, SubscriptionSink, SubscriptionMessage};
+	/// use jsonrpsee_types::ErrorObjectOwned;
+	///
+	/// let mut ctx = RpcModule::new(99_usize);
+	/// ctx.register_subscription("sub", "notif_name", "unsub", |params, pending, ctx| {
+	///
+	///     // The params are parsed outside the async block below to avoid cloning the bytes.
+	///     let val = match params.one::<usize>() {
+	///         Ok(val) => val,
+	///         Err(e) => {
+	///             // If the subscription has not been "accepted" then
+	///             // the return value will be "ignored" as it's not
+	///             // allowed to send out any further notifications on
+	///             // on the subscription.
+	///             tokio::spawn(pending.reject(ErrorObjectOwned::from(e)));
+	///             return Ok(());
+	///         }
+	///     };
+	///
+	///     tokio::spawn(async move {
+	///         // Mark the subscription is accepted after the params has been parsed successful.
+	///         // This is actually responds the underlying RPC method call and may fail if the
+	///         // connection is closed.
+	///         let sink = pending.accept().await.unwrap();
+	///         let sum = val + (*ctx);
+	///
+	///         let msg = SubscriptionMessage::from_json(&sum).unwrap();
+	///
+	///         // This fails only if the connection is closed
+	///         sink.send(msg).await.unwrap();
+	///     });
+	///
+	///     Ok(())
+	/// });
+	/// ```
+	///
+	pub fn register_subscription<R, F>(
+		&mut self,
+		subscribe_method_name: &'static str,
+		notif_method_name: &'static str,
+		unsubscribe_method_name: &'static str,
+		callback: F,
+	) -> Result<&mut MethodCallback, Error>
+	where
+		Context: Send + Sync + 'static,
+		F: (Fn(Params, PendingSubscriptionSink, Arc<Context>) -> R) + Send + Sync + Clone + 'static,
+		R: IntoSubscriptionCloseResponse + Send + 'static,
+	{
+		if subscribe_method_name == unsubscribe_method_name {
```

Comment:
nit: Since the beginning of this is similar to `register_async_subscription` would it be possible to move some code into a shared function? Maybe we could have a `internal_verify_and_unsubscribe` which does the initial validation and registers the unsubscribe?

Reply:
sure, I just want know what you think about the API but unfortunately a breaking change :(

Reply:
Yep, that makes sense! From my perspective, this gives a nice boost in terms of performance, users that may value more ergonomics still the old method 👍